source/tests/system/nidigital_driver_api_tests.cpp

```diff
@@ -0,0 +1,335 @@
+  {
+    ::grpc::ClientContext context;
+    digital::GetAttributeViBooleanRequest request;
+    request.mutable_vi()->set_id(GetSessionId());
+    request.set_channel_name(channel_name);
+    request.set_attribute(attribute);
+    digital::GetAttributeViBooleanResponse response;
+
+    ::grpc::Status status = GetStub()->GetAttributeViBoolean(&context, request, &response);
+
+    EXPECT_TRUE(status.ok());
+    expect_api_success(response.status());
+    return response.value();
+  }
+
+  ViInt32 get_int32_attribute(const char* channel_name, digital::NiDigitalAttributes attribute)
+  {
+    ::grpc::ClientContext context;
+    digital::GetAttributeViInt32Request request;
+    request.mutable_vi()->set_id(GetSessionId());
+    request.set_channel_name(channel_name);
+    request.set_attribute(attribute);
+    digital::GetAttributeViInt32Response response;
+
+    ::grpc::Status status = GetStub()->GetAttributeViInt32(&context, request, &response);
+
+    EXPECT_TRUE(status.ok());
+    expect_api_success(response.status());
+    return response.value();
+  }
+
+  ViInt64 get_int64_attribute(const char* channel_name, digital::NiDigitalAttributes attribute)
+  {
+    ::grpc::ClientContext context;
+    digital::GetAttributeViInt64Request request;
+    request.mutable_vi()->set_id(GetSessionId());
+    request.set_channel_name(channel_name);
+    request.set_attribute(attribute);
+    digital::GetAttributeViInt64Response response;
+
+    ::grpc::Status status = GetStub()->GetAttributeViInt64(&context, request, &response);
+
+    EXPECT_TRUE(status.ok());
+    expect_api_success(response.status());
+    return response.value();
+  }
+
+  std::string get_string_attribute(const char* channel_name, digital::NiDigitalAttributes attribute)
+  {
+    ::grpc::ClientContext context;
+    digital::GetAttributeViStringRequest request;
+    request.mutable_vi()->set_id(GetSessionId());
+    request.set_channel_name(channel_name);
+    request.set_attribute(attribute);
+    digital::GetAttributeViStringResponse response;
+
+    ::grpc::Status status = GetStub()->GetAttributeViString(&context, request, &response);
+
+    EXPECT_TRUE(status.ok());
+    expect_api_success(response.status());
+    return response.value();
+  }
+
+ private:
+  std::shared_ptr<::grpc::Channel> channel_;
+  std::unique_ptr<::nidevice_grpc::Session> driver_session_;
+  std::unique_ptr<digital::NiDigital::Stub> nidigital_stub_;
+  std::unique_ptr<nidevice_grpc::SessionRepository> session_repository_;
+  std::unique_ptr<digital::NiDigitalLibrary> nidigital_library_;
+  std::unique_ptr<digital::NiDigitalService> nidigital_service_;
+  std::unique_ptr<::grpc::Server> server_;
+};
+
+TEST_F(NiDigitalDriverApiTest, SelfTest_SelfTestCompletesSuccessfully)
+{
+  ::grpc::ClientContext context;
+  digital::SelfTestRequest request;
+  request.mutable_vi()->set_id(GetSessionId());
+  digital::SelfTestResponse response;
+
```

Comment:
SelfTest, Reset and SelfCalibrate test at the bottom of this file don't really have a "setup" section. The test name as well has only two components
1. Action -> SelfTest
2. Validation ->SelfTestCompletesSuccessfully
So this empty line shouldn't be there. Pattern we follow is, number of newlines in test body should be same as number of "_" in test name.

Fix all 3 tests mentioned above.

Reply:
fixed this. i had earlier got the pattern wrong. now i understood it